### PR TITLE
[SPARK-45554][PYTHON] Introduce flexible parameter to `assertSchemaEqual`

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to add three new parameters to the `assertSchemaEqual`: `ignoreNullable`, `ignoreColumnOrder` and `ignoreColumnName` to provide users with more flexibility in schema testing.

### Why are the changes needed?

To enhance the utility of `assertSchemaEqual` by accommodating various common schema comparison scenarios that users might encounter, without necessitating manual adjustments or workarounds.

### Does this PR introduce _any_ user-facing change?

Yes. `assertDataFrameEqual` now have the option to use the five new parameters:
<!DOCTYPE html>

Parameter | Type | Comment
-- | -- | --
ignoreNullable | Boolean [optional] | Specifies whether a column’s nullable property is included when checking for schema equality.</br></br> When set to True (default), the nullable property of the columns being compared is not taken into account and the columns will be considered equal even if they have different nullable settings.</br></br>When set to False, columns are considered equal only if they have the same nullable setting.
ignoreColumnOrder | Boolean [optional] | Specifies whether to compare columns in the order they appear in the DataFrames or by column name.</br></br> When set to False (default), columns are compared in the order they appear in the DataFrames.</br></br> When set to True, a column in the expected DataFrame is compared to the column with the same name in the actual DataFrame. </br></br>ignoreColumnOrder cannot be set to True if ignoreColumnNames is also set to True.
ignoreColumnName | Boolean [optional] | Specifies whether to fail the initial schema equality check if the column names in the two DataFrames are different.</br></br> When set to False (default), column names are checked and the function fails if they are different.</br></br> When set to True, the function will succeed even if column names are different. Column data types are compared for columns in the order they appear in the DataFrames.</br></br> ignoreColumnNames cannot be set to True if ignoreColumnOrder is also set to True.

### How was this patch tested?

Added usage examples into doctest for each parameter.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by itholic** [comment]
This also CI passed. Gentle reminder for @HyukjinKwon, also cc @ueshin @zhengruifeng .

File: python/pyspark/testing/utils.py [hunk 1]
```diff
@@ -328,6 +376,26 @@ def assertSchemaEqual(actual: StructType, expected: StructType):
     ?                               ^^                               ^^^^^
     + StructType([StructField('id', StringType(), True), StructField('amount', LongType(), True)])
     ?                               ^^^^                              ++++ ^
+
+    Different schemas (ignoring column order)
```
Comment (allisonwang-db) [inline python/pyspark/testing/utils.py:380]:
```suggestion
    Compare two schemas ignoring the column order
```

File: python/pyspark/testing/utils.py [hunk 2]
```diff
@@ -328,6 +376,26 @@ def assertSchemaEqual(actual: StructType, expected: StructType):
...
+
+    >>> s1 = StructType(
+    ...     [StructField("a", IntegerType(), True), StructField("b", DoubleType(), True)]
+    ... )
+    >>> s2 = StructType(
+    ...     [StructField("b", DoubleType(), True), StructField("a", IntegerType(), True)]
+    ... )
+    >>> assertSchemaEqual(s1, s2, ignoreColumnOrder=True)
+
+    Different schemas (ignoring column names)
```
Comment (allisonwang-db) [inline python/pyspark/testing/utils.py:390]:
```suggestion
    Compare two schemas ignoring the column names
```

File: python/pyspark/testing/utils.py [hunk 3]
```diff
@@ -311,11 +342,28 @@ def assertSchemaEqual(actual: StructType, expected: StructType):
 
     Examples
     --------
+    >>> from pyspark.pandas.utils import default_session
+    >>> spark = default_session()
```
Comment (allisonwang-db) [inline python/pyspark/testing/utils.py:346]: Why do we need this?

**Review by allisonwang-db — COMMENTED** [review]
These parameters will be super helpful!

File: python/pyspark/testing/utils.py (hunk 3 above)
Comment (itholic) [reply python/pyspark/testing/utils.py:346]: Oh, it's dummy line from my local testing. Will remove it. Thanks for catching this out!

File: python/pyspark/testing/utils.py (hunk 2 above)
Comment (itholic) [reply python/pyspark/testing/utils.py:390]: Applied the suggestions. Thanks!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**